src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi

```diff
@@ -327,7 +402,31 @@ cdef class _AioCall(GrpcCallWrapper):
         metadata_sent_observer()
 
         # Receives initial metadata.
-        initial_metadata_observer(
-            await _receive_initial_metadata(self,
-                                            self._loop),
+        self._set_initial_metadata(
+            await _receive_initial_metadata(self, self._loop)
         )
+
+
+cdef _AioCall new_AioCall(AioChannel channel, object deadline,
+                          bytes method, CallCredentials call_credentials):
+    """Instantiate a new _AioCall object.
+
+    Internally uses a raw initialization without using the traditional
```

Comment:
For future reference, here is the generated C++ code for the `__cinit__` currently on master:
<details>
<summary>__cinit__ Source</summary>

```C++
static int __pyx_pw_6cygrpc_8_AioCall_1__cinit__(PyObject *__pyx_v_self, PyObject *__pyx_args, PyObject *__pyx_kwds) {
  struct __pyx_obj_6cygrpc_AioChannel *__pyx_v_channel = 0;
  PyObject *__pyx_v_deadline = 0;
  PyObject *__pyx_v_method = 0;
  struct __pyx_obj_6cygrpc_CallCredentials *__pyx_v_call_credentials = 0;
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  __Pyx_RefNannySetupContext("__cinit__ (wrapper)", 0);
  {
    static PyObject **__pyx_pyargnames[] = {&__pyx_n_s_channel,&__pyx_n_s_deadline,&__pyx_n_s_method,&__pyx_n_s_call_credentials,0};
    PyObject* values[4] = {0,0,0,0};
    if (unlikely(__pyx_kwds)) {
      Py_ssize_t kw_args;
      const Py_ssize_t pos_args = PyTuple_GET_SIZE(__pyx_args);
      switch (pos_args) {
        case  4: values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
        CYTHON_FALLTHROUGH;
        case  3: values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
        CYTHON_FALLTHROUGH;
        case  2: values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
        CYTHON_FALLTHROUGH;
        case  1: values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
        CYTHON_FALLTHROUGH;
        case  0: break; 
        default: goto __pyx_L5_argtuple_error;
      }      
      kw_args = PyDict_Size(__pyx_kwds);
      switch (pos_args) {
        case  0:
        if (likely((values[0] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_channel)) != 0)) kw_args--;
        else goto __pyx_L5_argtuple_error;
        CYTHON_FALLTHROUGH;
        case  1:
        if (likely((values[1] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_deadline)) != 0)) kw_args--;
        else { 
          __Pyx_RaiseArgtupleInvalid("__cinit__", 1, 4, 4, 1); __PYX_ERR(33, 28, __pyx_L3_error)
        }      
        CYTHON_FALLTHROUGH;
        case  2:
        if (likely((values[2] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_method)) != 0)) kw_args--;
        else { 
          __Pyx_RaiseArgtupleInvalid("__cinit__", 1, 4, 4, 2); __PYX_ERR(33, 28, __pyx_L3_error)
        }      
        CYTHON_FALLTHROUGH;
        case  3:
        if (likely((values[3] = __Pyx_PyDict_GetItemStr(__pyx_kwds, __pyx_n_s_call_credentials)) != 0)) kw_args--;
        else { 
          __Pyx_RaiseArgtupleInvalid("__cinit__", 1, 4, 4, 3); __PYX_ERR(33, 28, __pyx_L3_error)
        }      
      }      
      if (unlikely(kw_args > 0)) {
        if (unlikely(__Pyx_ParseOptionalKeywords(__pyx_kwds, __pyx_pyargnames, 0, values, pos_args, "__cinit__") < 0)) __PYX_ERR(33, 28, __pyx_L3_error)
      }      
    } else if (PyTuple_GET_SIZE(__pyx_args) != 4) {
      goto __pyx_L5_argtuple_error;
    } else { 
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
      values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
      values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
    }
    __pyx_v_channel = ((struct __pyx_obj_6cygrpc_AioChannel *)values[0]);
    __pyx_v_deadline = values[1];
    __pyx_v_method = ((PyObject*)values[2]);
    __pyx_v_call_credentials = ((struct __pyx_obj_6cygrpc_CallCredentials *)values[3]);
  }
  goto __pyx_L4_argument_unpacking_done;
  __pyx_L5_argtuple_error:;
  __Pyx_RaiseArgtupleInvalid("__cinit__", 1, 4, 4, PyTuple_GET_SIZE(__pyx_args)); __PYX_ERR(33, 28, __pyx_L3_error)
  __pyx_L3_error:;
  __Pyx_AddTraceback("cygrpc._AioCall.__cinit__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __Pyx_RefNannyFinishContext();
  return -1;
  __pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_channel), __pyx_ptype_6cygrpc_AioChannel, 1, "channel", 0))) __PYX_ERR(33, 29, __pyx_L1_error)
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_method), (&PyBytes_Type), 1, "method", 1))) __PYX_ERR(33, 31, __pyx_L1_error)
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_call_credentials), __pyx_ptype_6cygrpc_CallCredentials, 1, "call_credentials", 0))) __PYX_ERR(33, 32, __pyx_L1_error)
  __pyx_r = __pyx_pf_6cygrpc_8_AioCall___cinit__(((struct __pyx_obj_6cygrpc__AioCall *)__pyx_v_self), __pyx_v_channel, __pyx_v_deadline, __pyx_v_method, __pyx_v_call_credentials);

  /* function exit code */
  goto __pyx_L0;
  __pyx_L1_error:;
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

static int __pyx_pf_6cygrpc_8_AioCall___cinit__(struct __pyx_obj_6cygrpc__AioCall *__pyx_v_self, struct __pyx_obj_6cygrpc_AioChannel *__pyx_v_channel, PyObject *__pyx_v_deadline, PyObject *__pyx_v_method, struct __pyx_obj_6cygrpc_CallCredentials *__pyx_v_call_credentials) {
  int __pyx_r;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  PyObject *__pyx_t_2 = NULL;
  PyObject *__pyx_t_3 = NULL;                         
  int __pyx_t_4;
  __Pyx_RefNannySetupContext("__cinit__", 0);
  
  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":33
 *                   bytes method,
 *                   CallCredentials call_credentials, /):
 *         self.call = NULL             # <<<<<<<<<<<<<<
 *         self._channel = channel               
 *         self._references = []
 */
  __pyx_v_self->__pyx_base.call = NULL;                                                                                                      

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":34
 *                   CallCredentials call_credentials, /):
 *         self.call = NULL
 *         self._channel = channel             # <<<<<<<<<<<<<<
 *         self._references = []
 *         self._loop = asyncio.get_event_loop()
 */
  __Pyx_INCREF(((PyObject *)__pyx_v_channel));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_channel));
  __Pyx_GOTREF(__pyx_v_self->_channel);
  __Pyx_DECREF(((PyObject *)__pyx_v_self->_channel));                                                                                        
  __pyx_v_self->_channel = __pyx_v_channel;
  
  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":35
 *         self.call = NULL
 *         self._channel = channel
 *         self._references = []             # <<<<<<<<<<<<<<
 *         self._loop = asyncio.get_event_loop()
 *         self._create_grpc_call(deadline, method, call_credentials)                                                                        
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(33, 35, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->_references);
  __Pyx_DECREF(__pyx_v_self->_references);
  __pyx_v_self->_references = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":36
 *         self._channel = channel
 *         self._references = []
 *         self._loop = asyncio.get_event_loop()             # <<<<<<<<<<<<<<
 *         self._create_grpc_call(deadline, method, call_credentials)
 *         self._is_locally_cancelled = False
 */
  __Pyx_GetModuleGlobalName(__pyx_t_2, __pyx_n_s_asyncio); if (unlikely(!__pyx_t_2)) __PYX_ERR(33, 36, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_2);
  __pyx_t_3 = __Pyx_PyObject_GetAttrStr(__pyx_t_2, __pyx_n_s_get_event_loop); if (unlikely(!__pyx_t_3)) __PYX_ERR(33, 36, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_3);
  __Pyx_DECREF(__pyx_t_2); __pyx_t_2 = 0;
  __pyx_t_2 = NULL;
  __pyx_t_4 = 0;
  if (CYTHON_UNPACK_METHODS && unlikely(PyMethod_Check(__pyx_t_3))) {
    __pyx_t_2 = PyMethod_GET_SELF(__pyx_t_3);
    if (likely(__pyx_t_2)) {
      PyObject* function = PyMethod_GET_FUNCTION(__pyx_t_3);
      __Pyx_INCREF(__pyx_t_2);
      __Pyx_INCREF(function);
      __Pyx_DECREF_SET(__pyx_t_3, function);
      __pyx_t_4 = 1;
    }
  }
  {
    PyObject *__pyx_callargs[1] = {__pyx_t_2, };
    __pyx_t_1 = __Pyx_PyObject_FastCall(__pyx_t_3, __pyx_callargs+1-__pyx_t_4, 0+__pyx_t_4);
    PyObject *__pyx_callargs[1] = {__pyx_t_2, };
    __pyx_t_1 = __Pyx_PyObject_FastCall(__pyx_t_3, __pyx_callargs+1-__pyx_t_4, 0+__pyx_t_4);
    __Pyx_XDECREF(__pyx_t_2); __pyx_t_2 = 0;
    if (unlikely(!__pyx_t_1)) __PYX_ERR(33, 36, __pyx_L1_error)
    __Pyx_GOTREF(__pyx_t_1);
    __Pyx_DECREF(__pyx_t_3); __pyx_t_3 = 0;
  }
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_self->_loop);
  __Pyx_DECREF(__pyx_v_self->_loop);
  __pyx_v_self->_loop = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":37
 *         self._references = []
 *         self._loop = asyncio.get_event_loop()
 *         self._create_grpc_call(deadline, method, call_credentials)             # <<<<<<<<<<<<<<
 *         self._is_locally_cancelled = False
 * 
 */
  ((struct __pyx_vtabstruct_6cygrpc__AioCall *)__pyx_v_self->__pyx_vtab)->_create_grpc_call(__pyx_v_self, __pyx_v_deadline, __pyx_v_method, __pyx_v_call_credentials); if (unlikely(PyErr_Occurred())) __PYX_ERR(33, 37, __pyx_L1_error)

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":38
 *         self._loop = asyncio.get_event_loop()
 *         self._create_grpc_call(deadline, method, call_credentials)
 *         self._is_locally_cancelled = False             # <<<<<<<<<<<<<<
 * 
 *     def __dealloc__(self):
 */
  __pyx_v_self->_is_locally_cancelled = 0;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":28
 * cdef class _AioCall(GrpcCallWrapper):
 * 
 *     def __cinit__(self,             # <<<<<<<<<<<<<<
 *                   AioChannel channel,
 *                   object deadline,
 */

  /* function exit code */
  __pyx_r = 0;
  goto __pyx_L0;
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_XDECREF(__pyx_t_2);
  __Pyx_XDECREF(__pyx_t_3);
  __Pyx_AddTraceback("cygrpc._AioCall.__cinit__", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = -1;
  __pyx_L0:;
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}

```

</details>

Meanwhile, here's the generated code for this new `new_AioCall` function.

<details><summary>new_AioCall Source</summary>

```C++
static struct __pyx_obj_6cygrpc__AioCall *__pyx_f_6cygrpc_new_AioCall(struct __pyx_obj_6cygrpc_AioChannel *__pyx_v_channel, PyObject *__pyx_v_deadline, PyObject *__pyx_v_method, struct __pyx_obj_6cygrpc_CallCredentials *__pyx_v_call_credentials) {
  struct __pyx_obj_6cygrpc__AioCall *__pyx_v_aio_call = 0;
  struct __pyx_obj_6cygrpc__AioCall *__pyx_r = NULL;
  __Pyx_RefNannyDeclarations
  PyObject *__pyx_t_1 = NULL;
  __Pyx_RefNannySetupContext("new_AioCall", 0);

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":421
 *     """
 *     cdef _AioCall aio_call
 *     aio_call = _AioCall.__new__(_AioCall)             # <<<<<<<<<<<<<<
 *     aio_call.call = NULL
 *     aio_call._channel = channel
 */
  __pyx_t_1 = ((PyObject *)__pyx_tp_new_6cygrpc__AioCall(((PyTypeObject *)__pyx_ptype_6cygrpc__AioCall), __pyx_empty_tuple, NULL)); if (unlikely(!__pyx_t_1)) __PYX_ERR(33, 421, __pyx_L1_error)
  __Pyx_GOTREF(((PyObject *)__pyx_t_1));
  __pyx_v_aio_call = ((struct __pyx_obj_6cygrpc__AioCall *)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":422
 *     cdef _AioCall aio_call
 *     aio_call = _AioCall.__new__(_AioCall)
 *     aio_call.call = NULL             # <<<<<<<<<<<<<<
 *     aio_call._channel = channel
 *     aio_call._loop = channel.loop
 */
  __pyx_v_aio_call->__pyx_base.call = NULL;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":423
 *     aio_call = _AioCall.__new__(_AioCall)
 *     aio_call.call = NULL
 *     aio_call._channel = channel             # <<<<<<<<<<<<<<
 *     aio_call._loop = channel.loop
 *     aio_call._references = []
 */
  __Pyx_INCREF(((PyObject *)__pyx_v_channel));
  __Pyx_GIVEREF(((PyObject *)__pyx_v_channel));
  __Pyx_GOTREF(__pyx_v_aio_call->_channel);
  __Pyx_DECREF(((PyObject *)__pyx_v_aio_call->_channel));
  __pyx_v_aio_call->_channel = __pyx_v_channel;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":424
 *     aio_call.call = NULL
 *     aio_call._channel = channel
 *     aio_call._loop = channel.loop             # <<<<<<<<<<<<<<
 *     aio_call._references = []
 *     aio_call._status = None
 */
  __pyx_t_1 = __pyx_v_channel->loop;
  __Pyx_INCREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_aio_call->_loop);
  __Pyx_DECREF(__pyx_v_aio_call->_loop);
  __pyx_v_aio_call->_loop = __pyx_t_1;
  __pyx_t_1 = 0;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":425
 *     aio_call._channel = channel
 *     aio_call._loop = channel.loop
 *     aio_call._references = []             # <<<<<<<<<<<<<<
 *     aio_call._status = None
 *     aio_call._initial_metadata = None
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(33, 425, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_aio_call->_references);
  __Pyx_DECREF(__pyx_v_aio_call->_references);
  __pyx_v_aio_call->_references = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":426
 *     aio_call._loop = channel.loop
 *     aio_call._references = []
 *     aio_call._status = None             # <<<<<<<<<<<<<<
 *     aio_call._initial_metadata = None
 *     aio_call._waiters_status = []
 */
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_aio_call->_status);
  __Pyx_DECREF(__pyx_v_aio_call->_status);
  __pyx_v_aio_call->_status = Py_None;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":427
 *     aio_call._references = []
 *     aio_call._status = None
 *     aio_call._initial_metadata = None             # <<<<<<<<<<<<<<
 *     aio_call._waiters_status = []
 *     aio_call._waiters_initial_metadata = []
 */
  __Pyx_INCREF(Py_None);
  __Pyx_GIVEREF(Py_None);
  __Pyx_GOTREF(__pyx_v_aio_call->_initial_metadata);
  __Pyx_DECREF(__pyx_v_aio_call->_initial_metadata);
  __pyx_v_aio_call->_initial_metadata = Py_None;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":428
 *     aio_call._status = None
 *     aio_call._initial_metadata = None
 *     aio_call._waiters_status = []             # <<<<<<<<<<<<<<
 *     aio_call._waiters_initial_metadata = []
 *     aio_call._create_grpc_call(deadline, method, call_credentials)
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(33, 428, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_aio_call->_waiters_status);
  __Pyx_DECREF(__pyx_v_aio_call->_waiters_status);
  __pyx_v_aio_call->_waiters_status = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":429
 *     aio_call._initial_metadata = None
 *     aio_call._waiters_status = []
 *     aio_call._waiters_initial_metadata = []             # <<<<<<<<<<<<<<
 *     aio_call._create_grpc_call(deadline, method, call_credentials)
 *     aio_call._is_locally_cancelled = False
 */
  __pyx_t_1 = PyList_New(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(33, 429, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __Pyx_GIVEREF(__pyx_t_1);
  __Pyx_GOTREF(__pyx_v_aio_call->_waiters_initial_metadata);
  __Pyx_DECREF(__pyx_v_aio_call->_waiters_initial_metadata);
  __pyx_v_aio_call->_waiters_initial_metadata = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":430
 *     aio_call._waiters_status = []
 *     aio_call._waiters_initial_metadata = []
 *     aio_call._create_grpc_call(deadline, method, call_credentials)             # <<<<<<<<<<<<<<
 *     aio_call._is_locally_cancelled = False
 *     return aio_call
 */
  ((struct __pyx_vtabstruct_6cygrpc__AioCall *)__pyx_v_aio_call->__pyx_vtab)->_create_grpc_call(__pyx_v_aio_call, __pyx_v_deadline, __pyx_v_method, __pyx_v_call_credentials); if (unlikely(PyErr_Occurred())) __PYX_ERR(33, 430, __pyx_L1_error)

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":431
 *     aio_call._waiters_initial_metadata = []
 *     aio_call._create_grpc_call(deadline, method, call_credentials)
 *     aio_call._is_locally_cancelled = False             # <<<<<<<<<<<<<<
 *     return aio_call
*/
  __pyx_v_aio_call->_is_locally_cancelled = 0;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":432
 *     aio_call._create_grpc_call(deadline, method, call_credentials)
 *     aio_call._is_locally_cancelled = False
 *     return aio_call             # <<<<<<<<<<<<<<
 */
  __Pyx_XDECREF(((PyObject *)__pyx_r));
  __Pyx_INCREF(((PyObject *)__pyx_v_aio_call));
  __pyx_r = __pyx_v_aio_call;
  goto __pyx_L0;

  /* "src/python/grpcio/grpc/_cython/_cygrpc/aio/call.pyx.pxi":410
 * 
 * 
 * cdef _AioCall new_AioCall(AioChannel channel, object deadline,             # <<<<<<<<<<<<<<
 *                           bytes method, CallCredentials call_credentials):
 *     """Instantiate a new _AioCall object.
 */

  /* function exit code */
  __pyx_L1_error:;
  __Pyx_XDECREF(__pyx_t_1);
  __Pyx_AddTraceback("cygrpc.new_AioCall", __pyx_clineno, __pyx_lineno, __pyx_filename);
  __pyx_r = 0;
  __pyx_L0:;
  __Pyx_XDECREF((PyObject *)__pyx_v_aio_call);
  __Pyx_XGIVEREF((PyObject *)__pyx_r);
  __Pyx_RefNannyFinishContext();
  return __pyx_r;
}
```

</details>

Some takeaways from my comparison of the two:
 - The `__cinit__`-based version has a lot of code for pathways that we will never take, like parsing keyword arguments. We could remove those branches by taking advantage of [keyword-only arguments on a newer version of Cython](https://github.com/cython/cython/pull/2949). But those paths are guarded with `unlikely` macros, meaning they're probably not super impactful.
 - `self._loop = asyncio.get_event_loop()` was changed to `aio_call._loop = channel.loop`. This eliminated the only actual Python function call (as far as I can tell). This seems to be the only change on the codepath that we actually go down.

@pfreixes Would you mind running another benchmark with `channel.loop` changed to `asyncio.get_event_loop`? I'm wondering if the change from `__cinit__` to C-style constructor was actually what led to the performance improvement or if it was mostly the removal of the Python call. If it's not the cause of the performance regression, I'd like to keep the encapsulation provided by `__cinit__`.

Reply:
Maybe I was a bit too eager on that, yes the optimization that you can get by not having the default constructor could be negligible or unnoticeable.  That the only stuff that will be circumvented would be the arguments unpacking:

```cpp
      values[0] = PyTuple_GET_ITEM(__pyx_args, 0);
      values[1] = PyTuple_GET_ITEM(__pyx_args, 1);
      values[2] = PyTuple_GET_ITEM(__pyx_args, 2);
      values[3] = PyTuple_GET_ITEM(__pyx_args, 3);
```

And later on this other part:

```cpp
__pyx_L4_argument_unpacking_done:;
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_channel), __pyx_ptype_6cygrpc_AioChannel, 1, "channel", 0))) __PYX_ERR(33, 29, __pyx_L1_error)
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_method), (&PyBytes_Type), 1, "method", 1))) __PYX_ERR(33, 31, __pyx_L1_error)
  if (unlikely(!__Pyx_ArgTypeTest(((PyObject *)__pyx_v_call_credentials), __pyx_ptype_6cygrpc_CallCredentials, 1, "call_credentials", 0))) __PYX_ERR(33, 32, __pyx_L1_error)
  __pyx_r = __pyx_pf_6cygrpc_8_AioCall___cinit__(((struct __pyx_obj_6cygrpc__AioCall *)__pyx_v_self), __pyx_v_channel, __pyx_v_deadline, __pyx_v_method, __pyx_v_call_credentials);
```

Almost for sure that most of the optimization in the construction time would come from the fact that we had removed the call to the `get_event_loop`, but since we were not explicitly allowing the _AioCall instantiation at Python level I had the feeling that this could become a "quick-win".

I'll be back with some numbers comparing current code with an without the calling to the `get_event_loop`.

Reply:
Here we go, I've executed three tests running the benchmark for 100 seconds - originally the benchmarks that I ran the benchmarks took only 10 seconds, I've wanted to run them again for a longer time..

The results are narrowed to the `_AioChannel.call` [1] function since it is the last Python function that can be profiled with a python profiler, this function encapsulates basically the process of creating a new `AioCall`.

Avg time of `_AioChannel.call` with **master**
```python
>>> (5.116 / 1555389) * 1e6
3.2892093231982478
```
_(5.116 stands for total time accumulated for this funciton, 1555389 stands for number of requests made, 1e6 converting to microseconds)_ 

Avg of time of `AioChannel.call` with current PR
```python
>>> (4.406 / 1669214) * 1e6
2.639565687802762
```

Avg of time of `AioChannel.call` with current PR but replacing `loop = channel.loop` by `loop = asyncio.get_event_loop()`
```python
>>> (5.653 / 1694820) * 1e6
3.335457452708842
```

So you were right, the main benefit comes basically from removing the call to the `get_event_loop`.

Considering that the improvements - if there are - of moving from `__cinit__` to an ad-hoc method is imperceptible I would revert the usage of an ad-hoc method and use basically the traditionally `__cinit__`.

[1] https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/aio/channel.pyx.pxi#L96

Reply:
I've reverted the changes for using the traditional `__cinit__`.